gemd/units/impl.py

```diff
@@ -3,70 +3,132 @@
+        if token.type == OP:
+            if token.string not in _ALLOWED_OPERATORS:
+                raise UndefinedUnitError(f"Unrecognized operator: {token.string}")
+
+            if exponent_context or token.string in {"**", "^", ".", "-", "+"}:
+                # Exponents & unaries do not change context
+                blocks[-1].append(token)
+            elif token.string not in {}:
+                blocks.append([])
+
+            if token.string == '(':
+                operator_stack.append(token)
+            elif token.string == ')':
+                while operator_stack:  # don't worry about enforcing balance
+                    if operator_stack.pop().string == '(':
+                        break  # We found token's friend
+            elif token.string in {"**", "^"}:
+                operator_stack.append(token)
+                continue  # Break flow since next token is in exponent context
+        elif token.type == NAME:
+            if exponent_context or len(blocks[-1]) == 0 or blocks[-1][-1].type != NAME:
+                blocks[-1].append(token)
+            else:  # Break blocks for two units in a row
+                blocks.append([token])
+        elif token.type == NUMBER:
+            blocks[-1].append(token)
+        elif token.type == ERRORTOKEN:  # Keep non-legal Python symbols like °
+            blocks[-1].append(token)
+        # Drop other tokens, such as EOF
+
+        if len(operator_stack) > 0 and operator_stack[-1].string in {"**", "^"}:
+            operator_stack.pop()  # Exit context for this exponential
+
+    todo = []
+    blocks.pop(0)  # Leading term is not allowed to be a scaling factor
+    for block in blocks:
+        i_exp = next((i for i, t in enumerate(block) if t.string in {"**", "^"}), len(block))
+        i_name = next((i for i, t in enumerate(block) if t.type == NAME), None)
+        numbers = [(i, t.string) for i, t in enumerate(block) if t.type == NUMBER and i < i_exp]
+
+        if len(numbers) == 1:
+            position, value = numbers[0]
+            if i_exp != len(block):
+                raise ValueError(
+                    f"Scaling factors ({value}) with exponents are not supported ({input_string})"
+                )
+            if i_name is not None and i_name < position:
+                raise ValueError(f"Scaling factor ({value}) follows unit in {input_string}")
+            if float(value) != 1.0 and float(value) != 0.0:  # Don't create definitions for 0 or 1
+                block_string = input_string[block[0].start[1]:block[-1].end[1]]
+                if i_name is None:
+                    unit_string = None
+                else:
+                    unit_string = input_string[block[position + 1].start[1]:block[i_name].end[1]]
+                todo.append((block_string, value, unit_string))
+        elif len(numbers) > 1:
+            raise ValueError(
+                f"Replicate scaling factor ({[n[1] for n in numbers]}) in {input_string}"
+            )
```

Comment:
This evaluates which of the expression fragments actually contain scaling factors we want to substitute, and stashes those cases in a todo list.